src/github_runner_manager/reactive/process_manager.py

```diff
@@ -0,0 +1,140 @@
+#  Copyright 2024 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""Module for managing processes which spawn runners reactively."""
+import logging
+import os
+import shutil
+import signal
+
+# All commands run by subprocess are secure.
+import subprocess  # nosec
+from pathlib import Path
+
+from github_runner_manager.reactive.types_ import RunnerConfig
+from github_runner_manager.utilities import secure_run_subprocess
+
+logger = logging.getLogger(__name__)
+
+REACTIVE_RUNNER_LOG_DIR = Path("/var/log/reactive_runner")
+
+PYTHON_BIN = "/usr/bin/python3"
+REACTIVE_RUNNER_SCRIPT_MODULE = "github_runner_manager.reactive.runner"
+REACTIVE_RUNNER_CMD_LINE_PREFIX = f"{PYTHON_BIN} -m {REACTIVE_RUNNER_SCRIPT_MODULE}"
+PID_CMD_COLUMN_WIDTH = len(REACTIVE_RUNNER_CMD_LINE_PREFIX)
+PIDS_COMMAND_LINE = [
+    "ps",
+    "axo",
+    f"cmd:{PID_CMD_COLUMN_WIDTH},pid",
+    "--no-headers",
+    "--sort=-start_time",
+]
+UBUNTU_USER = "ubuntu"
+RUNNER_CONFIG_ENV_VAR = "RUNNER_CONFIG"
+
+
+class ReactiveRunnerError(Exception):
+    """Raised when a reactive runner error occurs."""
+
+
+def reconcile(quantity: int, runner_config: RunnerConfig) -> int:
+    """Reconcile the number of reactive runner processes.
+
+    Args:
+        quantity: The number of processes to spawn.
+        runner_config: The reactive runner configuration.
+
+    Raises a ReactiveRunnerError if the runner fails to spawn.
+
+    Returns:
+        The number of reactive runner processes spawned/killed.
+    """
+    pids = _get_pids()
+    current_quantity = len(pids)
+    logger.info("Current quantity of reactive runner processes: %s", current_quantity)
+    delta = quantity - current_quantity
+    if delta > 0:
+        logger.info("Will spawn %d new reactive runner process(es)", delta)
+        _setup_logging_for_processes()
+        for _ in range(delta):
+            _spawn_runner(runner_config)
+    elif delta < 0:
+        logger.info("Will kill %d process(es).", -delta)
+        for pid in pids[:-delta]:
+            logger.info("Killing reactive runner process with pid %s", pid)
+            try:
+                os.kill(pid, signal.SIGTERM)
```

Comment:
Could these processes be attached to idling/ busy runners?

Reply:
Given the processes don't stay alive whilst the job is active this should be fine.

I believe at some point we discussed having the script trigger reconciliation after the runner is done. From the other comment, can't we just check for whether the post job script was executed to determine if the job has finished? Perhaps that is a general optimisation that we make for all the runners where we spawn a little script after spawning a runner which checks for whether the post job script was executed and then triggers reconciliation

Reply:
We can put that on the backlog